### ISSUE TITLE
how to feed data to multiple GPU?

### ISSUE THREAD

**Issue by leoyang233** [NVIDIA/DALI#273]
Should I use one pipeline for each GPU? since there is a parameter "device_id" in the Pipeline class.

**Comment by JanuszL** [comment]
Hi,
That is the way indeed how it works now. We don't support and I don't know if we ever will support multiple GPUs in one pipeline. It will make things complicated to move data between GPUs.
You can check https://github.com/NVIDIA/DALI/tree/master/docs/examples/pytorch, https://github.com/NVIDIA/DALI/tree/master/docs/examples/tensorflow/demo or https://github.com/NVIDIA/DALI/blob/master/docs/examples/mxnet to see how multiple GPU execution could look like.

**Comment by leoyang233** [comment]
If different GPUs get data from disk independently, how do you define one epoch? 
I usually use one single reader from disk, and then dispatch the data to different GPUs. DALI uses a different way, and will this affect the training convergence rate?

**Comment by JanuszL** [comment]
Hi,
In case of DALI we split dataset in even way. So each GPU work in each epoch at a separate piece of it.
Our test for RN50 shows no difference regarding accuracy. Also, PyTorch DistributedSampler does a similar thing - it just splits dataset into even pieces that each GPU works separately.
For DALI epoch is data_set_size/GPU_num.
Please check docs/examples/pytorch or docs/examples/mxnet.

**Comment by Edwardmark** [comment]
@JanuszL the page cannot be found, would you please show a breif clean example of using multiple gpus? Thanks.

**Comment by JanuszL** [comment]
@Edwardmark 
Updated links:
- https://github.com/NVIDIA/DALI/tree/master/docs/examples/use_cases/mxnet
- https://github.com/NVIDIA/DALI/tree/master/docs/examples/use_cases/pytorch
- https://github.com/NVIDIA/DALI/tree/master/docs/examples/use_cases/tensorflow/resnet-n

What you need to do is:
- prepare a script in the same way as for single GPU - data loader + model
- set local rank to device_id (in the pipeline)
- set global rank to shard_id (in the reader)
- set world size to the number of shards (in the reader)
- run training using distributed data-parallel strategy where each process is responsible for training on one GPU

**Comment by Edwardmark** [comment]
@JanuszL Thanks, man.

**Comment by JanuszL** [comment]
You can also check [this guide](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/multigpu.html).
